### PR TITLE
Added repr string for Grouper and TimeGrouper

### PR DESCRIPTION
**Pull request by topper-123**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This is a new implementation of #17727. This is much simpler than the previous attempts. The outputs are now:

```python
>>> pd.Grouper(key='A')
Grouper(key='A')
>>> pd.Grouper(key='A', freq='2T')
TimeGrouper(key='A', freq=<2 * Minutes>, sort=True, closed='left', label='left', convention='e')
```

Tests still need to be written. I can do them when/if this is ok.

### CODE REVIEW

File: pandas/core/groupby.py [hunk 1]
```diff
@@ -233,6 +233,8 @@ class Grouper(object):
 
     >>> df.groupby(Grouper(level='date', freq='60s', axis=1))
     """
+    _attributes = (('key', None), ('level', None), ('freq', None),
```
Comment (jreback) [inline pandas/core/groupby.py]: again you don't need to specify the default, just list the attributes. The repr should include all the values, not just the defaults. Further this means you are defining defaults in 2 places, here and the init, not a good idea.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby.py (hunk 1 above)
Comment (topper-123) [reply pandas/core/groupby.py]: Ok, changed. Also added test for ``TimeGrouper.__repr__``. 

Should a note about this go in the 0.21.1 or 0.22 whatsnew.rst?

**Comment by topper-123** [comment]
The output is now:

```python
>>> pd.Grouper(key='A')
Grouper(key='A', level=None, freq=None, axis=0, sort=False)
>>> pd.Grouper(key='A', freq='H')
TimeGrouper(key='A', level=None, freq=<Hour>, sort=True, closed='left', label='left', how='mean', nperiods=None, axis=0, fill_method=None, limit=None, loffset=None, kind=None, convention='e', base=0)
```

``TimeGrouper´´ is quite verbose.

**Comment by jreback** [comment]
so what you can do is skip printing attributes that are None
makes it’s shorter w/o loss of generality

File: pandas/core/groupby.py [hunk 2]
```diff
@@ -333,6 +334,15 @@ def _set_grouper(self, obj, sort=False):
     def groups(self):
         return self.grouper.groups
 
+    def __repr__(self):
+        attrs_list = []
+        for attr_name in self._attributes:
```
Comment (jreback) [inline pandas/core/groupby.py]: can be a list comprehension

File: pandas/core/resample.py [hunk 3]
```diff
@@ -1025,6 +1025,9 @@ class TimeGrouper(Grouper):
     Use begin, end, nperiods to generate intervals that cannot be derived
     directly from the associated object
     """
+    _attributes = ('key', 'level', 'freq', 'sort', 'closed', 'label',
+                   'how', 'nperiods', 'axis', 'fill_method', 'limit',
```
Comment (jreback) [inline pandas/core/resample.py]: only should have public attributes

Comment (jreback) [reply pandas/core/resample.py]: iow you don’t need any special ones for TimeGrouper (which is publicly deprecated anyhow)

Comment (topper-123) [reply pandas/core/resample.py]: A ``TimeGrouper`` object is returned when calling ``Grouper `` with a ``freq`` arg, IOW it's only the top level location of ``TimeGrouper`` that is deprecated, right?

File: pandas/core/groupby.py (hunk 2 above)
Comment (topper-123) [reply pandas/core/groupby.py]: Ok.

File: pandas/core/resample.py (hunk 3 above)
Comment (jreback) [reply pandas/core/resample.py]: yes however you will never actually directly construct one
point being is the args are the same as Grouper (maybe a couple of extra which are prob not documented)

Comment (jreback) [reply pandas/core/resample.py]: can u add the missing arguments in the doc string of Grouper? eg  convention, base, loffset i think are only public ones

File: pandas/tests/test_resample.py [hunk 4]
```diff
@@ -3400,3 +3400,10 @@ def test_aggregate_with_nat(self):
 
             # if NaT is included, 'var', 'std', 'mean', 'first','last'
             # and 'nth' doesn't work yet
+
+    def test_repr(self):
+        result = repr(TimeGrouper(key='A', freq='H'))
+        expected = ("TimeGrouper(key='A', freq=<Hour>, sort=True, "
+                    "closed='left', label='left', how='mean', axis=0, "
+                    "convention='e', base=0)")
+        assert result == expected
```
Comment (jreback) [inline pandas/tests/test_resample.py]: a better test is to actually construct the object and compare to the original

File: pandas/core/resample.py (hunk 3 above)
Comment (topper-123) [reply pandas/core/resample.py]: Ah of course. Changed.

File: pandas/tests/test_resample.py (hunk 4 above)
Comment (topper-123) [reply pandas/tests/test_resample.py]: I don't understand you here.

File: pandas/core/resample.py (hunk 3 above)
Comment (topper-123) [reply pandas/core/resample.py]: Considering your last comment, it's actually possible to to:

```python
>>> pd.Grouper(key='A', freq='H', convention='s')
TimeGrouper(key='A', freq=<Hour>, axis=0, sort=True)
```

Note the output now misses ``convention='s'``. shouldn't ``TimeGrouper`` actually still have its own ``_attributes`` to account for this?

File: pandas/tests/test_resample.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/test_resample.py]: no, use the repr to actually construct a new object (that's the point of a repr), and compare to the original

File: pandas/core/resample.py (hunk 3 above)
Comment (jreback) [reply pandas/core/resample.py]: yeah it should take a copy of the super (Grouper) and add to it

File: doc/source/whatsnew/v0.21.1.txt [hunk 5]
```diff
@@ -22,7 +22,7 @@ Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
 - :meth:`Timestamp.timestamp` is now available in Python 2.7. (:issue:`17329`)
--
+- :class:`Grouper` and :class:`TimeGrouper` now have a friendly repr output.
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: issue number

**Review by jreback — CHANGES_REQUESTED** [review]
see my other comments

File: doc/source/whatsnew/v0.21.1.txt (hunk 5 above)
Comment (topper-123) [reply doc/source/whatsnew/v0.21.1.txt]: Ok, fixed.

File: pandas/core/resample.py (hunk 3 above)
Comment (topper-123) [reply pandas/core/resample.py]: Added ``_attributes``.

I don't know what ``base``, ``loffset`` and ``convention`` do, and can't quite figure it out from the code, so can't add explanations to the doc string, sorry. I added them, but without explanations. Do you have a text for each?

File: pandas/tests/test_resample.py (hunk 4 above)
Comment (topper-123) [reply pandas/tests/test_resample.py]: I still don't understand. Should the ``expected`` be wrapped in a ``repr()``.

**Comment by jreback** [comment]
can you rebase

File: pandas/core/resample.py [hunk 6]
```diff
@@ -1025,7 +1025,10 @@ class TimeGrouper(Grouper):
     Use begin, end, nperiods to generate intervals that cannot be derived
     directly from the associated object
     """
-
+    _attributes = Grouper._attributes + ('closed', 'label', 'how',
+                                         'nperiods', 'fill_method', 'limit',
```
Comment (jreback) [inline pandas/core/resample.py]: can you leave out ``nperiods``, also ``fill_method`` and ``limit`` which are deprecated.

Comment (jreback) [reply pandas/core/resample.py]: you can actually completely remove ``nperiods`` as this is non-functional.

File: pandas/tests/test_resample.py [hunk 7]
```diff
@@ -3416,3 +3416,10 @@ def test_aggregate_with_nat(self):
 
             # if NaT is included, 'var', 'std', 'mean', 'first','last'
             # and 'nth' doesn't work yet
+
+    def test_repr(self):
+        result = repr(TimeGrouper(key='A', freq='H'))
```
Comment (jreback) [inline pandas/tests/test_resample.py]: add the issue number

File: pandas/tests/test_resample.py [hunk 8]
```diff
@@ -3416,3 +3416,10 @@ def test_aggregate_with_nat(self):
...
+        expected = ("TimeGrouper(key='A', freq=<Hour>, axis=0, sort=True, "
+                    "closed='left', label='left', how='mean', "
```
Comment (jreback) [inline pandas/tests/test_resample.py]: can you add a similar test for Grouper (in test_groupby)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. some comments.

File: pandas/tests/test_resample.py (hunk 8 above)
Comment (topper-123) [reply pandas/tests/test_resample.py]: added

**Comment by topper-123** [comment]
green.

**Comment by jreback** [comment]
linting issue. otherwise lgtm. note that you can do

``make lint-diff`` to see linting in your patch, locally

**Review by jreback — APPROVED** [review]

File: pandas/core/resample.py [hunk 9]
```diff
@@ -1025,6 +1025,10 @@ class TimeGrouper(Grouper):
     Use begin, end, nperiods to generate intervals that cannot be derived
     directly from the associated object
     """
+    _attributes = Grouper._attributes + ('closed', 'label', 'how',
+                                         'nperiods', 'fill_method', 'limit',
```
Comment (jreback) [inline pandas/core/resample.py]: remove nperiods, fill_method, and limit

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/resample.py (hunk 9 above)
Comment (jreback) [reply pandas/core/resample.py]: and remove nperiods from the signature entirely, its is not used

Comment (topper-123) [reply pandas/core/resample.py]: Ok, but what's the benefit of not having fill_method and limit in _attributes?

Comment (jreback) [reply pandas/core/resample.py]: because they are deprecated (see the maybe_process_deprecations)

Comment (topper-123) [reply pandas/core/resample.py]: Also, there's this 
```
    Notes
    -----
    Use begin, end, nperiods to generate intervals that cannot be derived
    directly from the associated object
```

None of begin or end seem to exist, and if nperiods isn't used, I've just removed that note

Comment (jreback) [reply pandas/core/resample.py]: yeah remove that, looks bogus

**Comment by topper-123** [comment]
All green, @jreback

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**